File: Category/Graph/HLD.cpp
// 树链剖分（HLD）
//
// 1.  如果权值在边上的话，则先将权值压到深度大的点上，再建立线段树。
// 2.  结点x的子树对应的DFS序区间为[id[x], id[x] + siz[x] - 1]。
//
struct HLD {
    int n;
    vector<int> siz, top, dep, parent, in, out, seq;
    vector< vector<int>> g;
    int cur;

    HLD() {}
    HLD(int n) {
        init(n);
    }
    void init(int n) {
        this->n = n;
        siz.resize(n);
        top.resize(n);
        dep.resize(n);
        parent.resize(n);
        in.resize(n);
        out.resize(n);
        seq.resize(n);
        cur = 0;
        g.assign(n, {});
    }
    void addEdge(int u, int v) {
        g[u].push_back(v);
        g[v].push_back(u);
    }
    void work(int root = 0) {
        top[root] = root;
        dep[root] = 0;
        parent[root] = -1;
        dfs1(root);
        dfs2(root);
    }
    void dfs1(int u) {
        if (parent[u] != -1) {
            g[u].erase( find(g[u].begin(), g[u].end(), parent[u]));
        }

        siz[u] = 1;
        for (auto &v : g[u]) {
            parent[v] = u;
            dep[v] = dep[u] + 1;
            dfs1(v);
            siz[u] += siz[v];
            if (siz[v] > siz[g[u][0]]) {
                 swap(v, g[u][0]);
            }
        }
    }
    void dfs2(int u) {
        in[u] = cur++;
        seq[in[u]] = u;
        for (auto v : g[u]) {
            top[v] = v == g[u][0] ? top[u] : v;
            dfs2(v);
        }
        out[u] = cur;
    }
    int lca(int u, int v) {
        while (top[u] != top[v]) {
            if (dep[top[u]] > dep[top[v]]) {
                u = parent[top[u]];
            } else {
                v = parent[top[v]];
            }
        }
        return dep[u] < dep[v] ? u : v;
    }

    int dist(int u, int v) {
        return dep[u] + dep[v] - 2 * dep[lca(u, v)];
    }

    int jump(int u, int k) {
        if (dep[u] < k) {
            return -1;
        }

        int d = dep[u] - k;

        while (dep[top[u]] > d) {
            u = parent[top[u]];
        }

        return seq[in[u] - dep[u] + d];
    }

    bool isAncester(int u, int v) {
        return in[u] <= in[v] && in[v] < out[u];
    }

    int rootedParent(int u, int v) {
         swap(u, v);
        if (u == v) {
            return u;
        }
        if (!isAncester(u, v)) {
            return parent[u];
        }
        auto it =  upper_bound(g[u].begin(), g[u].end(), v, [&](int x, int y) {
            return in[x] < in[y];
        }) - 1;
        return *it;
    }

    int rootedSize(int u, int v) {
        if (u == v) {
            return n;
        }
        if (!isAncester(v, u)) {
            return siz[v];
        }
        return n - siz[rootedParent(u, v)];
    }

    int rootedLca(int a, int b, int c) {
        return lca(a, b) ^ lca(b, c) ^ lca(c, a);
    }
};


struct HLD {
    int n, idx;
    vector<vector<int>> ver;
    vector<int> siz, dep;
    vector<int> top, son, parent;
    vector<int> in, id, val;
    Segt segt;

    HLD(int n) {
        this->n = n;
        ver.resize(n + 1);
        siz.resize(n + 1);
        dep.resize(n + 1);

        top.resize(n + 1);
        son.resize(n + 1);
        parent.resize(n + 1);

        idx = 0;
        in.resize(n + 1);
        id.resize(n + 1);
        val.resize(n + 1);
    }
    void add(int x, int y) { // 建立双向边
        ver[x].push_back(y);
        ver[y].push_back(x);
    }
    void dfs1(int x) {
        siz[x] = 1;
        dep[x] = dep[parent[x]] + 1;
        for (auto y : ver[x]) {
            if (y == parent[x]) continue;
            parent[y] = x;
            dfs1(y);
            siz[x] += siz[y];
            if (siz[y] > siz[son[x]]) {
                son[x] = y;
            }
        }
    }
    void dfs2(int x, int up) {
        id[x] = ++idx;
        val[idx] = in[x]; // 建立编号
        top[x] = up;
        if (son[x]) dfs2(son[x], up);
        for (auto y : ver[x]) {
            if (y == parent[x] || y == son[x]) continue;
            dfs2(y, y);
        }
    }
    void modify(int l, int r, int val) { // 链上修改
        while (top[l] != top[r]) {
            if (dep[top[l]] < dep[top[r]]) {
                swap(l, r);
            }
            segt.modify(id[top[l]], id[l], val);
            l = parent[top[l]];
        }
        if (dep[l] > dep[r]) {
            swap(l, r);
        }
        segt.modify(id[l], id[r], val);
    }
    void modify(int root, int val) { // 子树修改
        segt.modify(id[root], id[root] + siz[root] - 1, val);
    }
    int ask(int l, int r) { // 链上查询
        int ans = 0;
        while (top[l] != top[r]) {
            if (dep[top[l]] < dep[top[r]]) {
                swap(l, r);
            }
            ans += segt.ask(id[top[l]], id[l]);
            l = parent[top[l]];
        }
        if (dep[l] > dep[r]) {
            swap(l, r);
        }
        return ans + segt.ask(id[l], id[r]);
    }
    int ask(int root) { // 子树查询
        return segt.ask(id[root], id[root] + siz[root] - 1);
    }
    void work(auto in, int root = 1) { // 在此初始化
        assert(in.size() == n + 1);
        this->in = in;
        dfs1(root);
        dfs2(root, root);
        segt.init(val); // 建立线段树
    }
    void work(int root = 1) { // 在此初始化
        dfs1(root);
        dfs2(root, root);
        segt.init(val); // 建立线段树
    }
};

File: Go/HLD.go
type HLD struct {
	n    int
	root int
	clk  int
	g    [][]int
	pa   []int
	siz  []int
	dep  []int
	top  []int
	in   []int
	out  []int
	seq  []int
}

func NewHLD(n int) *HLD {
	return &HLD{
		n:    n,
		root: 0,
		clk:  0,
		g:    make([][]int, n),
		pa:   make([]int, n),
		siz:  make([]int, n),
		dep:  make([]int, n),
		top:  make([]int, n),
		in:   make([]int, n),
		out:  make([]int, n),
		seq:  make([]int, 0),
	}
}

func (h *HLD) AddEdge(u, v int) { h.g[u] = append(h.g[u], v) }

func (h *HLD) SetRoot(r int) { h.root = r }

func (h *HLD) Work() {
	h.pa[h.root] = -1
	h.dep[h.root] = 0
	h.top[h.root] = h.root
	h.dfs1(h.root)
	h.dfs2(h.root)
}

func (h *HLD) dfs1(u int) {
	if h.pa[u] != -1 {
		for i, v := range h.g[u] {
			if v == h.pa[u] {
				h.g[u] = append(h.g[u][:i], h.g[u][i+1:]...)
				break
			}
		}
	}
	h.siz[u] = 1
	for _, v := range h.g[u] {
		h.pa[v] = u
		h.dep[v] = h.dep[u] + 1
		h.dfs1(v)
		h.siz[u] += h.siz[v]
		if h.siz[h.g[u][0]] < h.siz[v] {
			h.g[u][0], v = v, h.g[u][0]
		}
	}
}

func (h *HLD) dfs2(u int) {
	h.in[u] = h.clk
	h.clk++
	for _, v := range h.g[u] {
		if v == h.g[u][0] {
			h.top[v] = h.g[u][0]
		} else {
			h.top[v] = v
		}
		h.dfs2(v)
	}
	h.out[u] = h.clk
}

func (h *HLD) LCA(u, v int) int {
	for h.top[u] != h.top[v] {
		if h.dep[h.top[u]] < h.dep[h.top[v]] {
			u, v = v, u
		}
		u = h.pa[h.top[u]]
	}
	if h.dep[u] > h.dep[v] {
		u, v = v, u
	}
	return u
}

func (h *HLD) Dist(u, v int) int {
	return h.dep[u] + h.dep[v] - 2*h.dep[h.LCA(u, v)]
}

func (h *HLD) Jump(u, k int) int {
	if h.dep[u] < k {
		return -1
	}
	d := h.dep[u] - k
	for h.dep[h.top[u]] > d {
		u = h.pa[h.top[u]]
	}
	return h.seq[h.in[u]-h.dep[u]+d]
}

func (h *HLD) IsAncestor(u, v int) bool {
	return h.in[u] <= h.in[v] && h.in[v] < h.out[u]
}

func (h *HLD) RootedParent(u, v int) int {
	if u == v {
		return v
	}
	if !h.IsAncestor(v, u) {
		return h.pa[v]
	}
	idx := sort.Search(len(h.g[v]), func(i int) bool {
		return h.in[h.g[v][i]] > h.in[u]
	})
	return h.g[v][idx-1]
}

func (h *HLD) RootedSize(u, v int) int {
	if u == v {
		return h.n
	}
	if !h.IsAncestor(v, u) {
		return h.siz[v]
	}
	return h.n - h.siz[h.RootedParent(u, v)]
}

func (h *HLD) RootedLCA(a, b, c int) int {
	return h.LCA(a, b) ^ h.LCA(b, c) ^ h.LCA(c, a)
}
